classes/ThankYous/ThankYousRepository.php

```diff
@@ -0,0 +1,1121 @@
+	 * @param int[]|null $extranet_ids
+	 * @param bool       $allow_no_thanked
+	 * @return int[]
+	 */
+	public function GetRecentThankYousIds(?int $limit = null, ?int $offset = null, ?array $extranet_ids = null, bool $allow_no_thanked = true, ?array $date_range = null, ?array $thanked_user_ids = null, ?array $tag_ids = null)
+	{
+		$query = "
+			SELECT " . self::THANK_YOU_TABLE . ".id
+			FROM " . self::THANK_YOU_TABLE
+			. "	GROUP BY " . self::THANK_YOU_TABLE . ".id
+			ORDER BY " . self::THANK_YOU_TABLE . ".date_created DESC";
+
+		$query = $this->query_factory->GetQueryBuilder($query);
+
+		$query->setLimit($limit, $offset);
+
+		if (isset($date_range))
+		{
+			$this->QueryAddCreatedBetweenFilter($query, $date_range);
+		}
+
+		if (isset($extranet_ids) || isset($thanked_user_ids))
+		{
+			$query->AddJoin(self::THANK_YOU_TABLE, self::THANKED_USERS_TABLE, self::THANKED_USERS_TABLE, self::THANK_YOU_TABLE . ".id = " . self::THANKED_USERS_TABLE . ".thanks_id");
+		}
+
+		if (isset($thanked_user_ids))
+		{
+			$this->QueryAddThankedUserFilter($query, $thanked_user_ids);
+		}
+
+		if (isset($tag_ids))
+		{
+			$this->QueryJoinThankYouToTagged($query);
+			$this->QueryAddTagsFilter($query, $tag_ids);
+		}
+
+		if (isset($extranet_ids))
+		{
+			$query->AddJoin(self::THANKED_USERS_TABLE, self::USER_TABLE, self::USER_TABLE, self::THANKED_USERS_TABLE . ".user_id = " . self::USER_TABLE . ".id");
+			$this->QueryAddExtranetFilter($query, $extranet_ids, $allow_no_thanked);
+		}
+
+		$result = $this->db->query($query->GetQuery());
+
+		$thank_you_ids = [];
+		while ($row = $result->fetchArray())
+		{
+			$thank_you_ids[] = (int) $row['id'];
+		}
+
+		return $thank_you_ids;
+	}
+
+	/**
+	 * Given an array of Thank You IDs, returns an array of Thankeds, indexed by the Thank You's ID and then the Thanked's ID.
+	 *
+	 * @param int[] $ids
+	 * @return array[Thankable]
+	 */
+	public function GetThankYousThankedsByThankYouIds(array $ids)
+	{
+		if (count($ids) === 0)
+		{
+			return [];
+		}
+
+		$query_string = "SELECT id, item_id, object_type, object_id FROM " . self::THANKED_TABLE;
+
+		$query = $this->query_factory->GetQueryBuilder($query_string);
+
+		$query->AddWhereAndClause(self::THANKED_TABLE . ".item_id IN in:int:thank_you_ids", $ids);
+
+		$results = $this->db->query($query->GetQuery());
+
+		$thank_yous_thankeds = [];
+		$thankeds            = [];
+		while ($row = $results->fetchArray())
+		{
+			$id                    = (int) $row['id'];
+			$thank_you_id          = (int) $row['item_id'];
+			$owner_class_id        = (int) $row['object_type'];
+			$owner_classes_item_id = (int) $row['object_id'];
+
+			$thankeds[$id] = ['oclass' => $owner_class_id, 'id' => $owner_classes_item_id];
+
+			$thank_yous_thankeds[$thank_you_id][$id] = true;
+		}
+
+		try
+		{
+			$thankeds = $this->CreateThankablesFromOClasses($thankeds);
+		} catch (ThankYouOClass $exception)
+		{
+			throw new LogicException("Unexpected Exception thrown", null, $exception);
```

Comment:
Is this really a logic exception if the database values are incorrect?

Reply:
No you're right. This Exception should only ever occur if a type of Owner Class has been thanked that Thank Yous doesn't actually support, which is either a regression or a database anomally.

What about logging it and continuing :)?

Reply:
Continuing may cause an error in the next loop, cause it looks like it's expecting the values to be available no matter what.

I'd just remove the try catch.

Reply:
I looked at removing the try/catch, but I've realised that this code itself could probably be better as one rogue record will stop an entire list retriever from working, which seems pretty poor to me. I'm going to refactor it to be a bit more resilient and allow for it to create those which it can and log those that it can't.

Reply:
As discussed, fixed with logging and returning an empty array.

Fixed in 2803a3896e9ccccb15e88f25bdc6453b5bb03dc8